nflow-engine/src/main/java/io/nflow/engine/workflow/definition/AbstractWorkflowDefinition.java

```diff
@@ -101,23 +105,33 @@ public String getType() {
    * Return the initial state of the workflow.
    * @return Workflow state.
    */
-  public S getInitialState() {
+  public WorkflowState getInitialState() {
     return initialState;
   }
 
   /**
    * Return the generic error state of the workflow.
    * @return Workflow state.
    */
-  public S getErrorState() {
+  public WorkflowState getErrorState() {
     return errorState;
   }
 
   /**
    * Return all possible states of the workflow.
    * @return Set of workflow states.
    */
-  public abstract Set<S> getStates();
+  public Set<WorkflowState> getStates() {
+    return new HashSet<>(states);
+  }
+
+  /**
+   * Register a state to the workflow.
+   * @param state The state to be registered.
+   */
+  public void registerState(WorkflowState state) {
```

Comment:
If I understood correctly - earlier the state space was defined as enum values. This assured in compile-time that you cannot make non-existent state transitions.

After this PR there's a requirement that you explicitly register the states. `permit()` does this as a side-effect, but if you earlier had a state without incoming transitions (=no permits) then you need to add `registerState()` call for those orphan states when you migrate to 8.0.0. This could be mentioned in `CHANGELOG` and perhaps in the JavaDoc of this class too. E.g. `permit()` javadoc could say it does the registration (instead of "adding", whatever that means), and class-level javadoc that you need to register all the states now.

Reply:
AFAIK the compile-time safety is not really changed here. Firstly, when using `WorkflowInstanceService.updateWorkflowInstance`, the state was defined as `string`. Secondly, when using `NextAction.moveToState` etc., the state was defined as `WorkflowState`. So it was possible to make non-existent state transitions before as well. IMO it should be the responsibility of unit tests to ensure that state transitions are correct.

Regarding the new requirement to register all states, I updated the javadocs and changelog.

Reply:
Looks good! Meaning auto-scanning static states etc. I can't find ways to improve it further here.